okhttp-tests/src/test/java/okhttp3/DuplexTest.java

```diff
@@ -45,6 +47,115 @@
   private HandshakeCertificates handshakeCertificates = localhost();
   private OkHttpClient client = defaultClient();
 
+  @Ignore("pipe.fold() doesn't flush so this stalls")
+  @Test public void trueDuplexClientWritesFirst() throws IOException {
+    MockResponse mockResponse = new MockResponse();
+    mockResponse.clearHeaders();
+    MwsDuplexAccess.instance.setBody(mockResponse, new DuplexResponseBody() {
+      @Override public void onRequest(RecordedRequest request,
+          BufferedSource requestBody, BufferedSink responseBody) throws IOException {
+
+        assertEquals("request A", requestBody.readUtf8Line());
+        responseBody.writeUtf8("response B\n");
+        responseBody.flush();
+
+        assertEquals("request C", requestBody.readUtf8Line());
+        responseBody.writeUtf8("response D\n");
+        responseBody.flush();
+
+        assertEquals("request E", requestBody.readUtf8Line());
+        responseBody.writeUtf8("response F\n");
+        responseBody.flush();
+
+        assertNull(requestBody.readUtf8Line());
+        requestBody.close();
+        responseBody.close();
+      }
+    });
+    server.enqueue(mockResponse);
+    enableProtocol(Protocol.HTTP_2);
+
+    DuplexRequestBody duplexRequestBody = new DuplexRequestBody(null, 128);
+
+    Call call = client.newCall(new Request.Builder()
+        .url(server.url("/"))
+        .post(duplexRequestBody)
+        .build());
+
+    BufferedSink requestBody = duplexRequestBody.createSink(call);
+    requestBody.writeUtf8("request A\n");
+    requestBody.flush();
+
+    try (Response response = duplexRequestBody.awaitExecute()) {
+      BufferedSource responseBody = response.body().source();
+      assertEquals("response B", responseBody.readUtf8Line());
+
+      requestBody.writeUtf8("request C\n");
+      requestBody.flush();
+      assertEquals("response D", responseBody.readUtf8Line());
+
+      requestBody.writeUtf8("request E\n");
+      requestBody.flush();
+      assertEquals("response F", responseBody.readUtf8Line());
+
+      requestBody.close();
+      assertNull(responseBody.readUtf8Line());
+    }
+  }
+
+  @Test public void trueDuplexServerWritesFirst() throws IOException {
+    MockResponse mockResponse = new MockResponse();
+    mockResponse.clearHeaders();
+    MwsDuplexAccess.instance.setBody(mockResponse, new DuplexResponseBody() {
+      @Override public void onRequest(RecordedRequest request,
+          BufferedSource requestBody, BufferedSink responseBody) throws IOException {
+        responseBody.writeUtf8("response A\n");
+        responseBody.flush();
+        assertEquals("request B", requestBody.readUtf8Line());
+
+        responseBody.writeUtf8("response C\n");
+        responseBody.flush();
+        assertEquals("request D", requestBody.readUtf8Line());
+
+        responseBody.writeUtf8("response E\n");
+        responseBody.flush();
+        assertEquals("request F", requestBody.readUtf8Line());
+
+        responseBody.close();
```

Comment:
I thought asserting on non-test thread would be dangerous but you're safe because test don't pass if the response body doesn't close.

Reply:
In general yes! But in this case we’re protected by the fact that the other thread won’t complete normally unless this one does.